xbmc/utils/UnicodeConverter.cpp

```diff
@@ -0,0 +1,992 @@
+ *
+ * \return an iconv converter
+ */
+thread_local std::vector<CUnicodeConverterType> converters;
+
+CUnicodeConverterType& CUnicodeConverter::CInnerUCodeConverter::GetConverter(
+    StdConversionType converterType)
+{
+  if (converters.size() == 0)
+  {
+    converters.emplace_back(CUnicodeConverterType{UTF8_SOURCE, UTF32_CHARSET});
+    converters.emplace_back(
+        CUnicodeConverterType{UTF32_CHARSET, "UTF-8", CUnicodeConverter::m_Utf8CharMaxSize});
+    converters.emplace_back(CUnicodeConverterType{UTF32_CHARSET, WCHAR_CHARSET});
+    converters.emplace_back(CUnicodeConverterType{WCHAR_CHARSET, UTF32_CHARSET});
+    converters.emplace_back(
+        CUnicodeConverterType{WCHAR_CHARSET, "UTF-8", CUnicodeConverter::m_Utf8CharMaxSize});
+    converters.emplace_back(CUnicodeConverterType{UTF8_SOURCE, WCHAR_CHARSET});
+  }
+  CUnicodeConverterType& myConverter = converters[converterType];
+  return myConverter;
+}
+
+template<typename ST_IN_T_ARG,
+         typename ST_OUT_T_ARG,
+         typename ST_IN_T, //= typename enable_if_basic_string_view_t<ST_IN_T_ARG>,
+         typename ST_OUT_T, // = typename enable_if_basic_string_t<ST_OUT_T_ARG>,
+         typename IN_CHAR_T, // = typename ST_IN_T_ARG::value_type,
+         typename OUT_CHAR_T> //= typename ST_OUT_T_ARG::value_type>
+
+ST_OUT_T CUnicodeConverter::CInnerUCodeConverter::stdConvert(
+    const enum StdConversionType convertType,
+    const ST_IN_T input,
+    const bool failOnInvalidChar, /* = false */
+    const bool substituteBadCodepoints /* = true */)
+{
+  if (input.empty())
+    return ST_OUT_T{};
+
+  if (convertType < 0 || convertType >= NumberOfStdConversionTypes)
+    return ST_OUT_T{}; // Empty string on error
+
+  CUnicodeConverterType& converter =
+      CUnicodeConverter::CInnerUCodeConverter::GetConverter(convertType);
+
+  return convert<ST_IN_T, ST_OUT_T>(converter.PrepareConverter(),
+                                    converter.GetTargetSingleCharMaxLen(), input, failOnInvalidChar,
+                                    substituteBadCodepoints);
+}
+
+struct CharPtrPtrAdapter
+{
+  /* Ensures the char buffers fed to iconv are cast to the correct
+   * type for the platform.
+   *
+   * iconv may declare inbuf to be char** rather than const char** depending
+   * on platform and version, so provide a wrapper that handles both.
+   */
+  const char** pointer;
+  explicit CharPtrPtrAdapter(const char** p) : pointer(p) {}
+  operator char* *() { return const_cast<char**>(pointer); }
+  operator const char* *() { return pointer; }
+};
+
+template<typename ST_IN_T_ARG,
+         typename ST_OUT_T_ARG,
+         typename ST_IN_T, // enable_if_basic_string_view_t<ST_IN_T_ARG>,
+         typename ST_OUT_T, // = typename enable_if_basic_string_t<ST_OUT_T_ARG>,
+         typename IN_CHAR_T, // = typename ST_IN_T_ARG::value_type,
+         typename OUT_CHAR_T> //= typename ST_OUT_T_ARG::value_type>
+
+ST_OUT_T CUnicodeConverter::CInnerUCodeConverter::convert(
```

Comment:
this function needs a thorough review; it is security critical and implementing this is prone to edge case problems, off-by-ones etc.
unfortunately, I don't have enough time to do it at the moment. maybe someone else can pitch in?

Reply:
I assume that I should not mark this resolved since you are asking others to continue reviewing.